### PR TITLE
Stats Reset File

### PR DESCRIPTION
**Pull request by SandytheLizard**

### CODE REVIEW

File: tuxemon/states/combat/combat.py [hunk 1]
```diff
@@ -870,13 +874,22 @@ def trigger_player_run(self, player):
     def end_combat(self):
         """End the combat"""
         # TODO: End combat differently depending on winning or losing
+        old_stats_data = Technique.keep_old_stats(self)
         for player in self.active_players:
-            for mon in player.monsters:
+            for mon, i in zip(player.monsters, range(len(player.monsters))):
```
Comment (vnmabus) [inline tuxemon/states/combat/combat.py]: Use `enumerate` instead.

File: tuxemon/states/combat/combat.py [hunk 2]
```diff
@@ -870,13 +874,22 @@ def trigger_player_run(self, player):
...
                 mon.end_combat()
-
-        # clear action queue
+                statslugs = ['speed', 'hp', 'armour', 'melee', 'ranged', 'dodge']
+                for stat in range(len(self.old_stats_data[i])):
```
Comment (vnmabus) [inline tuxemon/states/combat/combat.py]: Use `enumerate` here too.

File: tuxemon/states/combat/combat.py [hunk 3]
```diff
@@ -163,7 +163,10 @@ def startup(self, **kwargs):
         self.show_combat_dialog()
         self.transition_phase("begin")
         self.task(partial(setattr, self, "phase", "ready"), 3)
-
+        self.old_stats_data = []
```
Comment (bitcraft) [inline tuxemon/states/combat/combat.py:166]: i don't think this should be here.  there is a concept of "a single source of truth", and this is spreading the "truth" around.  since techniques already have a copy of the old stats, then it doesn't need to be repeated here.  i would like to see this removed, since it is redundant.

Comment (bitcraft) [reply tuxemon/states/combat/combat.py:166]: another issues is active_players.  it may change in a battle if there are multiple trainers, which is supported.  so that being a list of values with no reference to monsters/trainers makes it impossible to make it correct if trainers change.

File: tuxemon/states/combat/combat.py [hunk 4]
```diff
@@ -663,6 +666,7 @@ def perform_action(self, user, technique, target=None):
         """
         technique.advance_round()
 
+
```
Comment (bitcraft) [inline tuxemon/states/combat/combat.py:669]: please avoid changing formatting unless it is more make it conforming.  PEP8 and other standards state the code should have no more than 2 newlines between code/comments.  adding this make 3 lines of whitespace, which is not conforming.

File: tuxemon/states/combat/combat.py [hunk 5]
```diff
@@ -870,13 +874,22 @@ def trigger_player_run(self, player):
...
+            for i, mon in enumerate(player.monsters):
                 mon.end_combat()
-
-        # clear action queue
+                statslugs = ['speed', 'hp', 'armour', 'melee', 'ranged', 'dodge']
+                for j, stat in enumerate(self.old_stats_data[i]):
+                    if  self.old_stats_data[i][stat] > 0:
+                            setattr(mon, statslugs[stat], self.old_stats_data[i][stat])
+                            print(getattr(mon, statslugs[stat]))
+                # clear action queue
```
Comment (bitcraft) [inline tuxemon/states/combat/combat.py:886]: remove this print statement

File: tuxemon/states/combat/combat.py [hunk 6]
```diff
@@ -870,13 +874,22 @@ def trigger_player_run(self, player):
...
         self._action_queue = list()
-
+        oldstatsdata = []
```
Comment (bitcraft) [inline tuxemon/states/combat/combat.py:888]: i don't understand what this is doing.

File: tuxemon/states/combat/combat.py [hunk 7]
```diff
@@ -870,13 +874,22 @@ def trigger_player_run(self, player):
     def end_combat(self):
         """End the combat"""
         # TODO: End combat differently depending on winning or losing
+        old_stats_data = Technique.keep_old_stats(self)
```
Comment (bitcraft) [inline tuxemon/states/combat/combat.py:877]: this code is hard to follow, and made worse because the old stats values are spread around the techniques and now the combat class.  it would be more clear if the monsters were checked for statuses, and if the status has changed their stats, then revert the stats using the old stats value from the technique.

**Review by bitcraft — CHANGES_REQUESTED** [review]

**Comment by xirsoi** [comment]
I have concerns about this implementation, please do not merge until I've had a chance to do some tests with it tonight.

File: tuxemon/states/combat/combat.py (hunk 7 above)
Comment (SandytheLizard) [reply tuxemon/states/combat/combat.py:877]: if you guys have a more efficient way of doing it thats fine, i admit this code turned out a bit hard to follow as i kept having issues and making changes. it does work though

Comment (bitcraft) [reply tuxemon/states/combat/combat.py:877]: this is going to be a source of problems in the future.  please add a comment so that future developers can narrow down issues in battles with multiple trainers and strange stats changes.
```
# The values here are set in the technique and do not contain reference to trainers.  This code needs to be updated to support multiple trainers.
```

File: tuxemon/states/combat/combat.py (hunk 6 above)
Comment (bitcraft) [reply tuxemon/states/combat/combat.py:888]: can you explain what this is doing, or remove it if its not doing anything?

File: tuxemon/states/combat/combat.py (hunk 3 above)
Comment (bitcraft) [reply tuxemon/states/combat/combat.py:166]: i believe this is redundant now?

**Review by bitcraft — CHANGES_REQUESTED** [review]

**Comment by kerizane** [comment]
Why is the old stats stored on the technique class? Why not just attach it to the monster itself.?

**Review by kerizane — CHANGES_REQUESTED** [review]
Why are the old stats stored on the technique class? Why not just attach it to the monster itself?
